design-proposals/external-monitoring.md

```diff
@@ -0,0 +1,173 @@
+would be agreed upon in this PR.
+
+## Motivation
+
+### Development speed, and the ability to propose new ideas quickly
+
+Monitoring structure aims to be similar across different repositories. Therefore,
+there is a major overhead of aligning with different components. For example, let's say that we decide
+that similar alerts need to pop up from all of our Kubevirt repos, let's say that it's a `component_X_is_unhealthy`.
+In other words, we would have `component_virt_handler_is_unhealthy`, `component_virt_controller_is_unhealthy` etc for
+kubevirt/kubevirt, and similar alert for kubevirt/CDI, kubevirt/HCO, etc. This not only causes a burden of implementing
+these alerts separately in every repo, but also demands alignment between every different repo. Developers of every
+repository can differ on their opinions, leading to a huge effort of aligning between them all. Eventually, this leads
+to a situation in which it's very difficult to promote changes and new ideas that require a great amount of discussions
+and alignment.
+
+Once we have something like `kubevirt/monitoring` with a completely decoupled and autonomous from the operator, the
+need for alignment would be reduced significantly. There will still be a need to agree upon which data the operator
+exposes and how, but the monitoring logic itself will be decided in one single place.
+
+### Decoupling monitoring code from operator code
+
+Today, monitoring code resides within the operator's code. This approach has many disadvantages (in addition to the
+alignment problem mentioned above).
+
+To begin with, adding monitoring code to the heart of an operator's logic is unscalable from a maintainability perspective.
+As times goes on, operators are expected to add more and more metrics and alerts. This can cause a situation in which
+the portion of code that's dedicated to monitoring logic becomes too large and visible. As an example, in Kubevirt,
+we report to Prometheus about VMs changing phases, starting / ending migrations, etc. That code is coupled to the heart
+of our most sensitive logic, making our code less cohesive and more messy.
+
+Moreover, this design significantly increases the development burden when adding / changing monitoring logic. Since the
+monitoring code resides at the heart of a sensitive flow, code reviews will have to be strict. Often, people that are
+expert on migration flows aren't familiar with monitoring areas and vice-versa. This often leads to a very large development
+effort in order to merge new PRs.
+
+### Enhanced security: monitor publicly available data only
+
+Today we don't have a clear definition of which data is private and internal, and which data is public.
+
+Let's take an example: the amount of times that virt-controller had reconciled a VM. Currently, this is a completely
+internal data that isn't exposed. On the other hand, monitoring code can report it as metrics to Prometheus. So, is this
+public or private data?
```

Comment:
@iholder-redhat I'm not sure what you mean here. A metric is exposed if it can benefit either the operator for automation, the developers for regression tests and other purposes and for users that need to understand what is the state of their env and if they have issues they get alerts that are based on metrics.
There is no "private/internal" vs. "public" data. We add metrics if they are needed and bring value to any of the stakeholders.

Reply:
Let me give you an example.
Let's say that we have an operator that runs a certain CRD. For the example's sake, let's say it's a Human CRD. Now let's say that this Human object has the following fields as part of its status: **age** and **date of birth**.

Now let's say that we want to monitor the Human's location. With how things currently work, we'd probably place monitoring logic in the Human controller which notifies Prometheus about each location change for this Human. In other words, the monitoring code now exposes the human's location to the world, although this location **is not part of Human's status**. So is the location supposed to be exposed, or a hidden implementation detail?

---------

This is very similar to what's happening today in Kubevirt.
As an example - do we expose the number of reconciliations our controller have done? Only to Prometheus. The same goes for VM's dirty rate, for example. This design is problematic. Every piece of data that we're dealing with needs to either be completely open and reachable to the outside world, or kept secretly as an implementation detail.

Following this design will enforce us to think about which data is exposed or not and would eventually help to secure the privacy of our objects. With the current design it's highly probable that we'd accidentally expose something we're not supposed to.